Config parameter and interface enhancements

cdjackson opened the pull request:
This is an early PR for some changes to the config interface. It's not fully tested, but it's working with no obvious problems - I wanted to put it out there early for comment...

This adds a few features that are basically targeted towards devices with a lot of configuration - ie zwave being my focus - but I'm sure will also be useful for smaller bindings. Currently there are devices in the zwave database with 30+ configuration parameters, plus there will be additional parameters added by the binding in order to control parts of the device that don't need to be defined (eg wakeup intervals) as they are dependant on the command classes supported by the individual devices.

## Overview of Changes
1. It changes the XML config reader to use the new ConfigParameterDescriptionBuilder. This just makes good sense I think as it isolates the parameter definition, meaning changes can be made without breaking the XML reader (or any other reader).
2. I've added 3 new attributes to the configuration description.
 1. An 'advanced' flag. This allows the user interface to filter out parameters that are marked as 'advanced' - effectively highlighting parameters that are most useful.
 2. Added a 'limitToOptions' flag. This is used to say whether or not the value returned must be in the options list. The point here is that sometimes you want to have an options list of 'most used values' (or whatever) where the values are kind of hints, and other times you really want to limit to the listed values. A use case I have in zwave is a parameter may have a valid range of between (say) 50 and 100, but you can specify -1 to disable. I want to be able to set an option with value -1, and then let the user either select this, or a number of their choosing.
 3. Added a group attribute. This is a larger change, and is my attempt to solve the issues we've discussed on the ESH forum a while back about linking configuration parameters together. I feel this is a good compromise - it provides a good deal of definition, but doesn't break the current interface (and yes, I've checked that the PaperUI still works! :) ).

## Parameter Groups
Most of the changes are associated with the last point above - parameter groups. The config interface I currently have in OH1 for zwave is hierarchical, so it's easy to group things together. It was designed this way to be flexible, but in reality for a single device, and assuming the current interface in ESH, 1 level is more than enough.

So, the proposed solution is to add an (optional) group id string to each parameter so that the UI can (optionally) group the parameters.

In addition to the group id in each parameter, there is a parameterGroups list that has been added to the ConfigDescription. This has 4 members -:

1. groupid: to allow the UI to link parameters.
2. context: to allow some sort of context information - the UI might for example use it to display an icon. Ideally we'd agree a defined set (similar to the Categories) and potentially also limit them (enum) so the UI knows what they mean (which I also think should be done with Categories).
3. label: the group name for us mortals to read.
4. description: to let people know what it's all about.

These get exposed through the REST UI with the addition of the parameterGroups object in the thing. This makes it totally transparent - the UI can use it, or ignore it.

The above provides a concrete and consistent set of parameters and groups, all specified into a single object. They can be dynamically updated through the ConfigDescriptionProvider interface if the binding desires.

I've not looked at adding this to the XML provider in any detail yet. It should be considered optional so maybe it's not needed (but it would be nice I think), but for complex bindings, I think it's essential to avoid the user ending up with a list of 50+ parameters in a long list.

## More stuff to do
There's a few more bits to do - more testing for starters... 

I would like to resolve the number of selections that can be made when 'multiple' is set ([see bug 465493](https://bugs.eclipse.org/bugs/show_bug.cgi?id=465493)). Some configurations need multiple selection, but each device only allows a certain number of entries in some lists - I think it's better to limit this at the UI that in the binding and leave the user wondering why he set 10 entries, but he's only got 5.  This is an easy change but we can discuss in the bug what option is best.

I've not implemented this in the XML - I've had a quick look and given a little time I'm sure I can work that out, so I can probably add the above features into the XML config provider if desired.

I'm happy to tweak this a bit if anyone has other ideas...

Signed-off-by: Chris Jackson <chris@cd-jackson.com>

dnobel commented:
Hi Chris,

Thanks for this PR.

Some general comments from my side:

> An 'advanced' flag. This allows the user interface to filter out parameters that are marked as 'advanced' - effectively highlighting parameters that are most useful.

Sounds useful

> Added a 'limitToOptions' flag. This is used to say whether or not the value returned must be in the options list. 

Actually our current definition of the options should already define that this is the list of valid options. I would expect any other value to be invalid. So the flag should be 'notLimitedToOptions'. What is your use case here?

> Added a group attribute. 

I agree that grouping parameters could help if there are bindings with dozens of parameter. The approach with having an optional groupId seems to be backward-compatible.

> In addition to the group id in each parameter, there is a parameterGroups list that has been added to the ConfigDescription. This has 4 members -:

I think an advanced flag for the whole group may also be useful.

> I've not looked at adding this to the XML provider in any detail yet. It should be considered optional so maybe it's not needed

We definitely should update the XML schema and parser, so that it can be specified in the XML. It should also be possible to implement it backward-compatible.

If we agree to this changes, the PR must be extended to also cover XML schema, parser, tests and documentation. But it is a good idea to first discuss the API and the approach before going too deep into that stuff.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescriptionParameter.java:
```diff
@@ -58,6 +59,9 @@
     private String name;
     private Type type;
+
+    private String group;
```
I would call the parameter groupName or groupId.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescriptionParameter.java:
```diff
@@ -281,6 +308,33 @@ public String getLabel() {
+     *
+     * @return a group for the configuration parameter (could be null or empty)
+     */
+    public String getGroup() {
```
According to the previous comment I would call it getGroupName or getGroupId.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ParameterGroup.java:
```diff
@@ -0,0 +1,90 @@
+ * @author Chris Jackson - Initial Contribution
+ *
+ */
+public class ParameterGroup {
```
According to the other class name I would call it 'ConfigDescriptionParamterGroup'.

cdjackson replied:
No problem with the above changes.
The reason I went with ParameterGroup is because there is also ParameterOption - I'm happy to change though (it was a tossup between the two anyway).

cdjackson commented:
> > Added a 'limitToOptions' flag. This is used to say whether or not the value returned must be in the options list.

> Actually our current definition of the options should already define that this is the list of valid options. I would expect any other value to be invalid. So the flag should be 'notLimitedToOptions'. What is your use case here?

That's fine - we can call it whatever we like - the point is if you're allowed to enter other values, or if it is limited to those in the list - I would like this to be configurable.
The immediate use case I have is the one I mentioned above. Some configurations might allow a range to be entered (say 50 to 100), but they might define other values like -1 for off, or 0 for always, or something. So, I want to be able to configure something along the following lines -:
* min=50, max=120, options=[off=-1, always=0]. The alternative is to either add all 72 options for every number, or to not use the options, and just use the range, but then min needs to be -1, and we end up with a big invalid range in the middle.

The other more general use case is to use it as a combo-box (which allows exactly this use case) - here we might provide some options, but the user can add something else if they wanted. An example here (maybe) is that in HABmin1 (which has combo boxes) I allow the user to type in a format string (%.1f) but I also have a dropdown with a bunch of predefined options (eg "Float - 1 decimal place") so that the user doesn't have to dig into the format syntax.

> The approach with having an optional groupId seems to be backward-compatible.
Yep - this was my #1 goal :) It has been tested already in HABmin (where I've added some of the extra features) and PaperUI which still works fine :)

> I think an advanced flag for the whole group may also be useful.
Agreed - I'll add this to the group.

>  But it is a good idea to first discuss the API and the approach before going too deep into that stuff.
Yep - this was also my thinking :) I'm happy to look at the XML/tests/etc, but don't want to waste my time if this was to be rejected (or changed significantly).

kaikreuzer commented:
All suggestions look good to me as well and make a lot of sense. So if nobody vetoes, feel free to jump in the the "still to do" stuff :-) (btw: thanks, you are doing terrific stuff @cdjackson!)

cdjackson commented:
How about the 'limitToOptions' flag becomes 'freeFormInput' or something similar. It's a not uncommon term (I think?) for a combobox type control where you have a list of options but are allowed to enter other values.
So, freeFormInput=true would tell the UI not to limit to the values in [options]?

dnobel commented:
I don´t like the term "freeFormInput", as it is somehow a UI related term. The example you gave with the integers assumes that min and max can be combined with options. I´m not sure if this will be heavily used. 

Your example is also not a free form input, because you don´t want to have a string value for example. You just want to have the following set [-1, 50-100]. So the limitation of the value can just not be specified detailed enough.

I would propose to skip this flag for the moment. We already discussed something like dynamic value lists, where the Binding developer can implement an interface like `DynamicValueProvider`, which can be called from the UI. This would allow to programmatically define a list of valid values at runtime. But I would definitely move this into another PR.

cdjackson commented:
Hi Dennis,
Ok, but what about the second use case I mentioned -:
An example here is that in HABmin1 I allow the user to type in a format string (%.1f) but I also have a dropdown with a bunch of predefined options (eg "Float - 1 decimal place") so that the user doesn't have to dig into the format syntax.

Maybe I didn't explain this well, but what it means is that we have the list of possible options (predefined formats) but an experienced user can also type their own value in. In this case it's not possible to define a list (well - it would be a huge list as we'd need to potentially have evry possible combination of options that the formatter provides...

Chris

cdjackson commented:
> Your example is also not a free form input, because you don´t want to have a string value for example. 

That's true, but again, you have already defined a pattern (RegEx) I think. So, by setting these options appropriately, I think it gives a good ability to configure the control as you see fit.

- The free-form input simply means the user can enter data from the keyboard (or not)
- The type allows the UI to know what to do with the data
- The pattern allows us to limit that input to numbers, phone numbers or any other regex defined pattern
- min / max allow us to limit these options when the type is INTEGER/DECIMAL
- The options allows us to specify specific values. If it's a string, then we probably don't do anything, but if it's an INTEGER/DECIMAL, then I would expect the UI to limit based on the min/max (if set).

My view on the max and min is that this is used to limit the manual input. If the configuration setting defines an option of Off=0, and min=10, max=20, then this should allow the user to type in a number from 10 to 20, but still select the Off option. If the configuration parameters, as defined in the XML (or in the binding if done dynamically) define these values, then I can't see why they should be acceptable. I see the min/max and options aren't mutually exclusive, but are used collectively to allow us to define the allowable values.

Does the fact that this functionality might not be "heavily used" really mean that it shouldn't be available? 

(sorry for the 2 part response - I had to pop off to a meeting)

dnobel commented:
> Does the fact that this functionality might not be "heavily used" really mean that it shouldn't be available?

Lets say my current personal motivation is to keep the feature set as small as possible, because the platform is growing fast and we need to make sure that we can guarantee stability, documentation and tutorials. And for sure each new feature brings new complexity. So we should just carefully choose the features, that are needed by many bindings/users. At least for the features that affect the Core APIs like Items, Things and also Config Descriptions (but this is my personal view - maybe @kaikreuzer has a different opinion)

But you convinced me, that we should be able to define if the option list is just a proposal or a list of valid values. Maybe we could take your first proposal 'limitToOptions', where the default value is true and it could be set to false, if the user can define any value?

kaikreuzer commented:
> Maybe we could take your first proposal 'limitToOptions', where the default value is true and it could be set to false, if the user can define any value?

I'd be fine with this approach, so feel free to go ahead with the implementation, Chris!

cdjackson commented:
Cheers guys.
Most of the coding is done, but I'm struggling a little with working out how the XML unmarshalling works - it's done a little differently than I've done before with xstream in both HABmin and the zwave binding (it's a bit more manual it seems)...

If I can get this resolved, then I should be done pretty quick - just a bit of debugging and add the tests (which I've now got running, so that should be quick) - I'm not sure if you can give any pointers?

What I've done is added the groups into the config-description - like so...

``` xml
<config-description>
            <group groupId="" />
            <group groupId="" />
            <parameter name="" />
            <parameter name="" />
</config-description>
```

My problem is working out how to split these out in ConfigDescriptionConverter. I'm not sure if the above structure is an issue and I should split the group out of config-description, but I think it's best placed here.

Any pointers?

Thanks
Chris

dnobel commented:
> I'm not sure if the above structure is an issue and I should split the group out of config-description,

No, I think the XML has the correct structure, but I would propose the following change:

```
<config-description>
            <group name="" />
            <group name="" />
            <parameter name="" />
            <parameter name="" />
</config-description>
```
> My problem is working out how to split these out in ConfigDescriptionConverter

Similar stuff should be done in the ThingTypeConverter, where you also have multiple notes.

cdjackson commented:
Thanks Dennis - I think I looked in the ThingTypeConverter, but I'll take another look tonight...

Regarding changing the groupId to name - I've reflected the names in the classes, and the reason it's groupId is because you asked me to change from name to groupId :) If I now change as you suggest above, the XML and the classes will be different. This in itself isn't a problem, but it just might be confusing since in the XML for the parameter, we have 'groupId', so now the groupId will be linked to name in the group. I think this is wrong and it should remain as groupId.

dnobel commented:
>  I've reflected the names in the classes

OK, but usually the XML differs a bit from the classes. One question is if the group should have an "id" or a "name". As have "name" as identifier for parameter, I just thought we should also have "name" for the group. For sure, the attribute for the group assignment should be "groupName" (or groupId).

So it could look like this:

```xml
<config-description>
            <group name="group1" />
            <group name="group2" />
            <parameter name="parameter1" groupName="group1" />
            <parameter name="parameter2" groupName="group2"/>
            <parameter name="parameterWithoutGroup"/>
</config-description>
```

What do you think?

cdjackson commented:
That's fine by me - I currently have groupId, but I'm happy to change this to groupName. However I would align the classes with the XML and call the ParameterGroup member 'name'. I think this is consistent as in the Parameter we also have 'name'.

I know the class and XML don't need to be the same, but I think it will avoid confusion.

dnobel commented:
Yes, if we use "name" for parameter Groups, I would expect that the ParameterGroup class has a member "name" and that the Parameter class gets a new member called "groupName" with the according getter "getGroupName()".

cdjackson commented:
I think this is now ready for review. The XML reader has been updated and tests added...

dnobel commented:
Hi Chris, 

thanks a lot.

From the first very rough look, it looks good. But I have seen, that you have "groupId" attribute in parameter tag in the schema. Can you change it to "groupName", as discussed previously? 

But before I can begin with a detailed review, can you please apply the code formatter for all files? There are a lot tabs in some files, e.g. the ConfigDescriptionParameter, which also leads to a difficult diff.

Regards

Dennis

cdjackson commented:
Hi Dennis, 
I’ll take a look - I swapped at least twice between groupId and groupName as our conversations changed, so I might have missed one (or two)…

Cheers
Chris

cdjackson commented:
Ok - hopefully this is formatted ok. I'm not sure what happened - my system still said it was using the ESH formatter, but it was sticking tabs in for some reason and I had to reload the format file! :(
I note that the XML file has changed formatting - I'm not sure if it just wasn't formatted before, or if something is wrong at my end still...

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescription.java:
```diff
@@ -26,11 +31,13 @@
 
     private URI uri;
     private List<ConfigDescriptionParameter> parameters;
+    private List<ConfigDescriptionParameterGroup> groups;
```
I would name that parameterGroups.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescription.java:
```diff
@@ -54,6 +61,22 @@ public ConfigDescription(URI uri) throws IllegalArgumentException {
+     *
+     * @param parameters the description of a concrete configuration parameter
+     *            (could be null or empty)
+     *
```
@param for groups is missing here

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescription.java:
```diff
@@ -94,9 +123,20 @@ public URI getURI() {
+     *
+     * @return the list of parameter groups parameter (not null, could be empty)
+     */
+    public List<ConfigDescriptionParameterGroup> getGroups() {
```
I would call the method getParameterGroups, that makes it more clear what kind of groups are meant.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescriptionParameter.java:
```diff
@@ -226,13 +263,22 @@ public Boolean isReadOnly() {
     }
 
     /**
+     * @return true if multiple selections of options are allowed, otherwise
```
Looks like copy and paste error.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescriptionParameter.java:
```diff
@@ -274,16 +322,46 @@ public String getDefault() {
     /**
+     * Returns a the group for this configuration parameter.
+     *
+     * @return a group for the configuration parameter (could be null or empty)
```
I would not allow empty value.

cdjackson replied:
I'm happy to change it, but I didn't modify this...

cdjackson replied:
Also, in the XSD this is optional - again - not my doing so I'd probably prefer to take this outside of this PR.

dnobel replied:
Well, it should be optional and can be null. I was just thinking wether it makes sense that the group name is an empty string. But let´s leave it as it is.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescriptionParameterGroup.java:
```diff
@@ -0,0 +1,105 @@
+ * A parameter group is used to group a number of parameters together so they can
+ * be displayed together in the UI (eg in a single tab).
+ * <p>
+ * A {@link ConfigDescriptionParameter} instance must also contain the groupId. It should be permissible to use the
```
should be named groupName now.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescriptionParameterGroup.java:
```diff
@@ -0,0 +1,105 @@
+     * Create a Parameter Group. A group is used by the user interface to display groups
+     * of parameters together.
+     *
+     * @param groupId
```
name

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigDescriptionParameterGroup.java:
```diff
@@ -0,0 +1,105 @@
+public class ConfigDescriptionParameterGroup {
+
+    private String name;
+    private String context;
```
Do we really need a context for groups? the contexts we want define for parameters are terms like "email", "password" and so on. I don´t think that they fit for groups.

cdjackson replied:
The context for parameters doesn't need to be the same as groups. I'd like to be able to provide some sort of context for this group to the user. In ZWave, there are a number of distinctive sets of data types that will be grouped together and I think it's useful to be able to indicate what 'type' they are.
Maybe context isn't the right term - I'm happy to change the name...

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/i18n/ConfigDescriptionGroupI18nUtil.java:
```diff
@@ -0,0 +1,50 @@
+ *
+ * @author Chris Jackson - Initial contribution
+ */
+public class ConfigDescriptionGroupI18nUtil {
```
Do you have an example for the default keys?

I would expect something like this: thing-type.config.`<binding-id>.<thing-type-id>.<group-name>.label` and `<binding-id>.<thing-type-id>.<group-name>.description`

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/i18n/ConfigDescriptionGroupI18nUtil.java:
```diff
@@ -0,0 +1,50 @@
+        this.i18nProvider = i18nProvider;
+    }
+
+    public String getGroupDescription(Bundle bundle, URI configDescriptionURI, String parameterName,
```
According to the previous comment the argument should be groupName and not parameterName.

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/i18n/ConfigDescriptionGroupI18nUtil.java:
```diff
@@ -0,0 +1,50 @@
+        return i18nProvider.getText(bundle, key, defaultDescription, locale);
+    }
+
+    public String getGroupLabel(Bundle bundle, URI configDescriptionURI, String parameterName, String defaultLabel,
```
According to the previous comment the argument should be groupName and not parameterName.

dnobel commented on bundles/config/org.eclipse.smarthome.config.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/config/config.xml:
```diff
@@ -8,13 +8,16 @@
 		</parameter>
 
 		<parameter name="username" type="text">
+			<groupName>user</groupName>
```
I would expect groupName to be an attribute and node a node. But lets discuss it.

dnobel commented on bundles/config/org.eclipse.smarthome.config.xml/config-description-1.0.0.xsd:
```diff
@@ -1,57 +1,80 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" 
```
Please check the formatter here. Looks like you changed a lot of spaces to tabs.

cdjackson replied:
Yes - I mentioned that I think at the end of the last commit. I'm using the ESH formatter - I do't know if it covers XML or not, or if there's anything special I need to do, but ESH formatter is selected...

cdjackson replied:
In Eclipse, is XML formatter specified somewhere? Currently I only have the java formatter and I can't find anywhere to add the XML...

dnobel replied:
Unfortunately it does not cover XML :(. So I suggest that you modify it manually without a formatter.

cdjackson replied:
Ok - wouldn’t we be better using the Eclipse inbuilt formatter - at least then can be made consistent if someone else changes it?

dnobel replied:
What is the inbuilt formatter? I think it depends on the XML editor you use.

cdjackson replied:
Yes :) I assumed though that Eclipse is being used (otherwise I guess the other Java formatter also won't work - or is it standard for all editors?)

dnobel commented on bundles/config/org.eclipse.smarthome.config.xml/config-description-1.0.0.xsd:
```diff
@@ -1,57 +1,80 @@
+				minOccurs="0" />
+			<xs:element name="filter" type="config-description:filterType"
+				minOccurs="0" />
+			<xs:element name="groupName" type="xs:string" minOccurs="0" />
```
As written in the previous comment I would make it an attribute.

dnobel commented on bundles/config/org.eclipse.smarthome.config.xml/src/main/java/org/eclipse/smarthome/config/xml/ConfigDescriptionConverter.java:
```diff
@@ -42,7 +46,6 @@ public ConfigDescriptionConverter() {
         this.attributeMapValidator = new ConverterAttributeMapValidator(new String[][] { { "uri", "false" } });
     }
 
-    @SuppressWarnings("unchecked")
```
Can it really be removed?

cdjackson replied:
Eclipse gives me a warning when it's there, and when I remove it, it doesn't... Does your environment show something different?

dnobel replied:
It gives me a warning if I remove it  - strange...

cdjackson replied:
![image](https://cloud.githubusercontent.com/assets/1041475/7457463/9b1fd0b2-f284-11e4-808e-8c5bc03908d9.png)

dnobel commented on docs/sources/architecture/configuration.md:
```diff
@@ -29,6 +29,13 @@ Configuration descriptions must be placed as XML file(s) (with the ending `.xml`
         http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
 
   <config-description uri="{binding|thing-type|bridge-type|channel-type|any_other}://bindingID:...">
+    <group name="String">
```
Must groups be specified before the parameters or after?

cdjackson replied:
Does it matter? I wasn't aware that xstream discriminated? To me, before makes sense...

dnobel replied:
No, it´s the schema. We defined sequences, which has a fixed order as far as I know.

cdjackson replied:
I just put one group before the parameters and one after and it passes the tests. I need to update Eclipse to use a local schema so I can validate what the schema thinks...

cdjackson replied:
Ok - you're right. It doesn't matter to the reader, so you can put it wherever you like and it reads in ok, but the schema wants it in a specific order... Do you prefer it before or after the parameters? To me, it seems best to put them first, but I'm not too bothered...

dnobel replied:
I don´t have a preference, so I´ll leave the decision up to you.

dnobel commented on docs/sources/architecture/configuration.md:
```diff
@@ -64,17 +71,34 @@ Configuration descriptions must be placed as XML file(s) (with the ending `.xml`
+Groups allow parameters to be grouped together into logical blocks so that the user can find the parameters they are looking for. A parameter can be placed into a group so that the UI knows how to display the information.
+<table>
+  <tr><td><b>Property</b></td><td><b>Description</b></td></tr>
+  <tr><td>group.name</td><td>The group name - this is used to link the parameters into the group, along with the groupName option in the parameter (mandatory).</td></tr>
```
group.name or groupName?

cdjackson replied:
group.name isn't it. It's an attribute called name, inside the group. I think this is consistent with the conventions used elsewhere in the document?

dnobel replied:
OK, agree!

dnobel commented on bundles/config/org.eclipse.smarthome.config.xml/config-description-1.0.0.xsd:
```diff
@@ -1,57 +1,80 @@
+		<xs:attribute name="multiple" type="xs:boolean" />
+	</xs:complexType>
+
+	<xs:complexType name="group">
```
I think this element should be named "parameterGroup".

dnobel commented:
Hi Chris, 

I´m done with code review. Looks great. We should just discuss the minor comments regarding the schema.

Moreover, can I also ask you
* to extend the i18n documentation about parameter groups and the i18n test (ConfigDescriptionI18nTest)? 
* to extend the thing type xml tests for embedded use of config description parameter groups (org.eclipse.smarthome.core.thing.xml.test).

Thanks!

Dennis

cdjackson commented:
Thanks for the review Dennis - much appreciated.

Re the schema - I don't have strong feelings either way. I'm always a little careful to add too many attributes as it can get messy and hard to read. That was my main driver to make it an element. If you strongly prefer it to be an attribute, I'll change it...

I'm not sure I understand what you mean about extending the test - what test do you mean? I think I've already got tests that test the parameter groups in a number of different ways. If there's a test case you'd like to see, can you be specific?

Cheers
Chris

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/thing/thing-types.xml:
```diff
@@ -22,6 +22,18 @@
 					access the API.
 				</description>
 			</parameter>
+			<config-group group-id="group1">
```
should be "groupName"

dnobel commented:
> If you strongly prefer it to be an attribute, I'll change it

I think it is more compliant to the stuff we already have (like thingTypeId) and so one. So I prefer it - but not strongly ;)

> I'm not sure I understand what you mean about extending the test - what test do you mean?

This one: org.eclipse.smarthome.core.thing.xml.test.ConfigDescriptionsTest. It tests, that so called "embedded" config descriptions can be read from the XML file. You adapted it, but you don´t check for parameter groups. The according XML file also looks wrong.

cdjackson commented:
> If you strongly prefer it to be an attribute, I'll change it
> 
> I think it is more compliant to the stuff we already have (like thingTypeId) and so one. So I prefer it - but not strongly ;)
> 
Ok - I’ll swap.

> I'm not sure I understand what you mean about extending the test - what test do you mean?
> 
> This one: org.eclipse.smarthome.core.thing.xml.test.ConfigDescriptionsTest. It tests, that so called "embedded" config descriptions can be read from the XML file. You adapted it, but you don´t check for parameter groups. The according XML file also looks wrong.
> 
Ah, yes, I didn’t add the tests here as I’d already done it in the config.core bundle, but you’re right, it needs a basic check at least in this bundle...

cdjackson commented:
Hey Dennis,
So, I think I've done everything :)
You might wanna take a look at the i18n stuff - clearly I missed most of that first time around so there've been a few changes there... I hope I've understood that correctly - I think so, and it seems to work ok, but please take a look :)

I haven't reformatted the XML file - my preference would be to stick with the TAB/spaces changes - ie to use the formatter so that any further changes at least can be run through the formatter. Up to you though...

Other than that I think I've implemented everything you've suggested and added the extra thing tests... Let me know what I've missed :)

Cheers
Chris

dnobel commented on docs/sources/architecture/internationalization.md:
````diff
@@ -116,11 +120,11 @@ channel-type.yahooweather.temperature.option.OPTION1 = Option Nummer 1
 ```
 
-So the key for referencing a label of a defined thing type is `thing-type.<binding-id>.<thing-type-id>.label`. A label of a channel can be referenced with `channel-type.<binding-id>.<channel-type-id>.label`. And finally the config description parameter key is `thing-type.config.<binding-id>.<thing-type-id>.<parameter-name>.label`.
+So the key for referencing a label of a defined thing type is `thing-type.<binding-id>.<thing-type-id>.label`. A label of a channel can be referenced with `channel-type.<binding-id>.<channel-type-id>.label`. And finally the config description parameter key is `thing-type.config.<binding-id>.<thing-type-id>.<parameter-name>.label` and the group parameter is `thing-type.configGroup.<binding-id>.<th ...
````
> and the group parameter is `thing-type.configGroup.<binding-id>.<thing-type-id>.<parameter-name>.label`.

I think the pattern is '<binding-id>.<thing-type-id>.<parameter-group-name>'

dnobel commented on bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/i18n/ConfigDescriptionGroupI18nUtil.java:
```diff
@@ -0,0 +1,50 @@
+
+    private String inferKey(URI configDescriptionURI, String groupName, String lastSegment) {
+        String uri = configDescriptionURI.getSchemeSpecificPart().replace(":", ".");
+        return configDescriptionURI.getScheme() + ".configGroup." + uri + "." + groupName + "." + lastSegment;
```
> ".configGroup."

I think the pattern should be: config and not configGroup. config is just a sub classifier for all config related texts.

So I would change this line to `configDescriptionURI.getScheme() + ".config." + uri + "." + groupName + "." + lastSegment;`

As long as now one defines a group with the same name as the parameter it should be unqiue.

cdjackson replied:
> As long as now one defines a group with the same name as the parameter it should be unqiue.

You mean **different** name right? If you defined a group with the same name as any parameter, then it will clash.  I'm happy to change this, it just seemed restricting to require group names and parameter names to be unique but I don't mind... If we keep the concept that all config related 'stuff' has to be unique, then it could start to become a problem if we add actions, and maybe some other types in future...

dnobel replied:
Yes I meant different. It will only clash for the i18n stuff, but the rest should work? But let´s continue the discussion at the bottom of this PR, otherwise we are mixing it up.

dnobel commented on bundles/config/org.eclipse.smarthome.config.xml.test/src/test/resources/test-bundle-pool/yahooweather.bundle/ESH-INF/i18n/yahooweather_de.properties:
```diff
@@ -6,4 +6,7 @@ config.config.Dummy.question.option.667931 = K
+config.config.Dummy.refresh.description = Spezifiziert das Aktualisierungsintervall in Sekunden
+
+config.configGroup.Dummy.group1.label = Group 1 German Label
+config.configGroup.Dummy.group1.description = Group 1 German Description
```
Regarding to my previous comment I would change it to `config.config.Dummy.group1.description`

dnobel commented on docs/sources/architecture/internationalization.md:
```diff
@@ -109,6 +109,10 @@ thing-type.yahooweather.weather.description = Stellt verschiedene Wetterdaten vo
 thing-type.config.yahooweather.weather.location.description = Ort der Wetterinformation.
 
+thing-type.configGroup.yahooweather.weather.group1.label = a label
+thing-type.configGroup.yahooweather.weather.group1.description = a description
```
Regarding to my previous commit I would change it to `thing-type.config.yahooweather.weather.group1.label = a label` and `thing-type.config.yahooweather.weather.group1.description = a description`

dnobel commented:
Hi Chris,

Thanks Again. I only have comment regarding the default key for the config group, I think it should be "config" instead of "configGroup" (see inline comments). The rest looks fine.

The last question is if we want to keep the context or rename it to something else. @kaikreuzer Any preference here?

Regards

Dennis

cdjackson commented:
Hi Dennis,
Thanks again. I’m not sure that I agree with the config name - the problem (maybe minor - I’m just not sure) is that if everything uses .config, then it means that parameters and groups, and in the future, actions, and anything else that might be added, all need to have unique names. Maybe this is ok, but it seems a little limiting since nothing else constrains this - just the i18n… I think it means that we can define a group, and a parameter each called xxx, and it will work fine until someone adds translations when it won’t work...

Chris

dnobel commented:
Well the key more or less adresses the XML path: `thing-type.binding.typeId.config.Dummy.group1.label` fits to `<thing-type><config-description><parameter>` and so one.

The only limitation I see is that groupNames and parameter names must be unique inside one config description. Not sure what you mean with action, but there will be only a problem if we put more stuff into the config description.

But even if we would to distinguish between parameters and groups we need to change the pattern to  `thing-type.binding.typeId.config.Dummy.group.groupName1.label` for groups and `thing-type.binding.typeId.config.Dummy.parameter.parameter1.label` for parameters. Which will break the current key scheme

cdjackson commented:
> Well the key more or less adresses the XML path:

More or less - but not the same, and that's the point. :) It's a constraint that only exists in the i18 files. I would personally be happier if it was the same as the XML path - not more, not less :)


> Not sure what you mean with action, but there will be only a problem if we put more stuff into the config description.

Agreed - that's my point... The 'action' reference was to the discussion on the Eclipse forum where there's a use case for actions to be added to the config interface (ie the ability to send a 'configuration command' to the binding to, for example, reset a bridge). I'm just worried that we start to set constraints that could be difficult for people to understand - it's not enforced anywhere, so there's nothing to stop someone (other than code review) creating a thing, and then at a later date, someone creating the translation that won't work...

dnobel commented:
OK, I agree. But I´m still not happy with the key you suggested. 

So what do you think about having this key: `thing-type.binding.typeId.config.Dummy.group.groupName1.label`. Later we can add stuff like `thing-type.binding.typeId.config.Dummy.action.actionName1.label`

cdjackson commented:
Agreed - that looks much nicer - thanks.

cdjackson commented:
So I think I'm done... Two things left to resolve I think -:
* The XML formatter - do you want me to go through and replace with spaces, or do we use the Eclipse XML formatter to ensure a common format. My preference is to use the formatter - at least we can make sure it's always consistent...
* The name of the _context_ attribute - does @kaikreuzer want it to be changed to a different name

...otherwise, I think I've covered all the comments and changes :) Let me know about the about 2 points, and anything I've missed, and I'll squash it...

dnobel commented:
The formatting of the file is OK for me. A suggestion for the context: Could you remove it for the this PR? As long as we do not define valid contexts it makes no sense, but we can discuss this topic separately and include it in another PR.

It looks like your PR can not be merged, probably because of some other merged PRs. If you agree to remove the context, I think you can squash your commits, adapt your PR to be merged and if Kai has no further concerns, we should be able to merge it.

kaikreuzer commented:
I am actually fine with leaving the context in there as they are. I just discussed with @rodneyrehm yesterday, that we will need predefined and reusable parameter groups, e.g. for "wifi" settings (SSID, pwd, etc), "authorization" (user, pwd). Similar as with the context for parameters, we should start with a small list and have it something optional, but over time we can work out what kind of groups occur more often and should as such be "standardized" by defining a context for them.

dnobel commented:
OK, then let´s keep the current state of the PR and you just need to adapt it, so that it can be merged.

cdjackson commented:
Ok - great - thanks. I'll merge the latest master into this branch tonight, resolve the conflicts and squash.

cdjackson commented:
Ok - I've merged master into this, and squashed, however I'm not completely sure it's squashed correctly (I hope so as I don't want to recreate the whole PR from scratch!).

It's still showing a merge conflict (the conflict I resolved earlier was in the configuration.md file), and the PR also has a bunch of other stuff in it from other merges. I'm using SourceTree's squash, so I hope it's picked it up ok...

Let me know how it looks......... (I fear all is not well - I'm not a fan of this squashing)

dnobel commented:
It does not look so good. Your branch is 30 commits behind the master and your commit contains a lot of other changes. 

You need to rebase the current master into your branch and then push it with the -force flag. But first you need a clean commit including only your changes. You don´t need to create another PR. With the -force flag you can override everything.

cdjackson commented:
Thanks. I think this looks ok now. I’ve not looked at every line, but it seems ok...

dnobel commented:
Hi Chris,

Thanks the PR is fine now. But I just had a final look and I want you to ask if you can make one last final change... :(

Currently the element in the xml is named "<parameterGroup>", but we have the convention, that xml tags are in dash-notation: "<parameter-group>", whereas attributes are camel-case. Could you please change it in the xsd, parser and documentation?

I think I have told you the wrong term, sorry for that. 

Regards

Dennis

cdjackson commented:
Hi Dennis,
I'm not sure what tags you mean - is there something missing in your response - the name is just specified as "" - maybe github has removed something?
Chris

dnobel commented:
Yes github removed it, sorry. It is about the `<parameterGroup>` tag. According to our convention it should be `<parameter-group>`.

cdjackson commented:
All done (I hope :) ).

dnobel commented:
Thanks Chris! Everything looks good now. @kaikreuzer From my perspective we can merge it. OK for you?

kaikreuzer commented on bundles/config/org.eclipse.smarthome.config.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/config/config.xml:
```diff
@@ -2,40 +2,65 @@
 	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
 	xmlns:config-description="http://eclipse.org/smarthome/schemas/config-description/v1.0.0"
-	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/config-description/v1.0.0 ../../../../../../../../org.eclipse.smarthome.config.xml/org.eclipse.smarthome.config-description.xsd ">
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/config-description/v1.0.0 http://eclipse.org/smarthome/schemas/config-description/v1.0.0">
```
This might actually fail validation in case you are binding a different version of the code than what is currently published to the Eclipse website. So having the relative location as it was before, definitely makes sense!

cdjackson commented:
The relative location caused a failure in my environment. It's much easier, and I think better, to use the Eclipse XML database - I don't know if this is stored in any file that can be put into the repo though?

kaikreuzer commented:
Well, the file is located in the repo https://github.com/eclipse/smarthome/blob/master/bundles/config/org.eclipse.smarthome.config.xml/config-description-1.0.0.xsd and the relative path refers to this location (which should work fine with the test execution). The problem is that if you try to build version 0.7.0 of the ESH code you might not be able to do so anymore, if the version on the ESH website is not compatible - so you would have lost the ability to reproduce the build of older versions, which is a bad thing.

cdjackson commented:
Yes, but do you think it's possible to use the Eclipse XML database? Do you know how that works under the hood - is it stored into the repo somewhere?

I think otherwise using the relative paths doesn't solve the problem with validating XML files since they still point to the old version of the XSD. However I also don't think that an invalid XML file will stop the build, or even the tests (does it? I certainly never noticed this while doing this development and I know the XML and XSD weren't always in sync).

kaikreuzer commented:
> Yes, but do you think it's possible to use the Eclipse XML database? Do you know how that works under the hood - is it stored into the repo somewhere?

There is nothing like an "Eclipse XML database" - I simply manually upload the XSDs to the webserver that hosts eclipse.org/smarthome.

>  since they still point to the old version of the XSD

No, they will always point to the version of your current code checkout - that's the whole point. If you checkout an older code version, you will also have an older (and thus compatible) XSD with it.

> However I also don't think that an invalid XML file will stop the build

I thought that there are some XSD validation tests, but I am not sure. So what exactly is the problem in your case? If there is no validation, it the the relative path should also not matter at all?

cdjackson commented:
> There is nothing like an "Eclipse XML database" - I simply manually upload the XSDs to the webserver that hosts eclipse.org/smarthome.

Maybe we're talking cross terms - I'm calling it a database, but the correct term is catalog, and I'm talking about the Eclipse IDE, not the Eclipse website.

In Eclipse IDE, click on PREFERENCES, then (from memory) XML, and then somewhere in there is a configuration for a local XML repository. This allows you to point the project to a local set of XML/XSD files that are stored within a project, or on the local file system. By using this, we can validate all XML files against the current version - using a relative path in one XML file solves the problem for that single XML file only - but it means we'd have to change every XML file in every binding if we do it your way (I think?).

https://wiki.eclipse.org/Using_the_XML_Catalog


> >since they still point to the old version of the XSD

> No, they will always point to the version of your current code checkout - that's the whole point

How does that work without the local catalog? If I edit a Thing XML file, it is pointing to the version on the Eclipse server - how does it know to use the local XSD that's embedded in the repo? The only way I know of to do this is to use the local catalog, but maybe I'm doing it wrong?


>  If there is no validation, it the the relative path should also not matter at all?
There's no validation in the tests is what I meant, but if you're editing an XML file, then in Eclipse you can click the VALIDATE button, and this will fail unless the XSD is correct. Of course you can ignore it - it doesn't stop if  compiling, but it also means that you end up with lots of red marks in the XML so you can't be sure it's right.
The way I resolve this is to use the Eclipse local catalog.

kaikreuzer commented:
We are talking here about a file that is used within a unit test - so I do not care about the Eclipse IDE and whether it correctly resolves the path and can validate it - what is important that it can be used during JUnit test execution (mainly when running the Maven build) and for this, the Eclipse XML catalog is irrelevant.

cdjackson commented:
Ok - I didn't pick up on the fact that it's the test, but as it currently stands, the tests pass with it referencing the old file, so I think that means it doesn't matter. I think the only validation is done manually - I've not actually checked this, but I've not had a test fail when the XSD has been wrong.

kaikreuzer commented:
So can we revert this line to the old relative location or is there any reason against it?

cdjackson commented:
I'll take a look at reverting it back - I don't think it's needed as it doesn't change the test, but I'll check it doesn't cause a problem.

cdjackson commented:
So, if I revert this back I get warnings in the XML that there is no grammar constraint. I'm happy to do that if you want it reverted. Either way makes no difference to the test as it's clearly not using this line at all (I can write complete rubbish in this line - I will still get the same warning, and the tests will still pass).

kaikreuzer commented:
Hm, this is unexpected to me, but if it really does not make any difference what is in there, then sorry for bothering and I will finally merge the PR!

marcelrv commented:
@cdjackson 
Hi Chris,

Did the advance parameter finally fully make it in this PR /xsd. ?
I saw it in the documentation but I'm not fully clear if it is an element or a parameter (from the doc it seems a parameter, from the code it seems a element)

cdjackson commented:
Hi Marcel,
Both advanced configs (i.e. both groups and parameters) are elements - you’re right - one of them is incorrect in the doc… I’ll sort this tomorrow…

Chris

> On 13 May 2015, at 23:00, Marcel <notifications@github.com> wrote:
> 
> @cdjackson <https://github.com/cdjackson> 
> Hi Chris,
> 
> Did the advance parameter finally fully make it in this PR /xsd. ?
> I saw it in the documentation but I'm not fully clear if it is an element or a parameter (from the doc it seems a parameter, from the code it seems a element)
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/eclipse/smarthome/pull/272#issuecomment-101829047>.
>

cdjackson commented:
Nope! I'll take a look when I get home early this afternoon. It looks like some sort of UTF-8 encoding mess to me.
